File: app/src/main/java/com/veyndan/paper/reddit/api/reddit/network/interceptor/AccessTokenInterceptor.java
package com.veyndan.paper.reddit.api.reddit.network.interceptor;

import com.veyndan.paper.reddit.api.reddit.network.AccessToken;
import com.veyndan.paper.reddit.api.reddit.network.AuthenticationService;
import com.veyndan.paper.reddit.api.reddit.network.Credentials;

import java.io.IOException;

import io.reactivex.Single;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public final class AccessTokenInterceptor implements Interceptor {

    private final AuthenticationService authenticationService;
    private final Credentials credentials;

    private AccessToken accessTokenCache = AccessToken.EXPIRED_ACCESS_TOKEN;

    public AccessTokenInterceptor(final AuthenticationService authenticationService,
                                  final Credentials credentials) {
        this.authenticationService = authenticationService;
        this.credentials = credentials;
    }

    @Override
    public Response intercept(final Chain chain) throws IOException {
        final Request unauthorizedRequest = chain.request();

        final Request authorizedRequest = Single.concat(accessTokenCache(), accessTokenNetwork())
                .filter(accessToken -> !accessToken.isExpired())
                .firstElement()
                .map(accessToken -> unauthorizedRequest.newBuilder()
                        .header("Authorization", "Bearer " + accessToken.getAccessToken())
                        .build())
                .blockingGet();

        return chain.proceed(authorizedRequest);
    }

    private Single<AccessToken> accessTokenCache() {
        return Single.just(accessTokenCache);
    }

    private Single<AccessToken> accessTokenNetwork() {
        final Single<AccessToken> single = authenticationService.getAccessToken(
                "password", credentials.getUsername(), credentials.getPassword())
                .map(retrofit2.Response::body);

        // Save access token from network into the cache.
        return single.doOnSuccess(accessToken -> accessTokenCache = accessToken);
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/api/reddit/network/interceptor/AccessTokenInterceptor.kt
package com.veyndan.paper.reddit.api.reddit.network.interceptor

import com.veyndan.paper.reddit.api.reddit.network.AccessToken
import com.veyndan.paper.reddit.api.reddit.network.AuthenticationService
import com.veyndan.paper.reddit.api.reddit.network.Credentials
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.schedulers.Timed
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response

class AccessTokenInterceptor(private val authenticationService: AuthenticationService,
                             private val credentials: Credentials) : Interceptor {

    private var accessTokenCache: Single<Timed<AccessToken>> = Observable.just(AccessToken.EXPIRED_ACCESS_TOKEN)
            .timestamp()
            .singleOrError()

    override fun intercept(chain: Interceptor.Chain): Response {
        val unauthorizedRequest: Request = chain.request()

        val authorizedRequest: Request = Single.concat(accessTokenCache, accessTokenNetwork())
                .filter { timed -> timed.time() + timed.value().expiresIn() >= System.currentTimeMillis() }
                .firstElement()
                .map { timed ->
                    unauthorizedRequest.newBuilder()
                            .header("Authorization", "Bearer ${timed.value().accessToken}")
                            .build()
                }
                .blockingGet()

        return chain.proceed(authorizedRequest)
    }

    private fun accessTokenNetwork(): Single<Timed<AccessToken>> {
        val single: Single<Timed<AccessToken>> = authenticationService.getAccessToken(
                "password", credentials.username, credentials.password)
                .map { it.body()!! }
                .toObservable()
                .timestamp()
                .singleOrError()

        // Save access token from network into the cache.
        return single.doOnSuccess { accessToken -> accessTokenCache = Single.just(accessToken) }
    }
}
